policy/engine.go

```diff
@@ -5,35 +5,109 @@ import (
 	"github.com/open-policy-agent/opa/storage"
 	"github.com/open-policy-agent/opa/topdown"
+	"github.com/open-policy-agent/opa/version"
 )
 
-// Engine represents the policy engine
+// Engine represents the policy engine.
 type Engine struct {
-	Compiler *ast.Compiler
-	Store    storage.Store
-	Trace    bool
+	result   *loader.Result
+	compiler *ast.Compiler
+	store    storage.Store
+	tracing  bool
 }
 
-// NewEngine returns a new instatiated Engine
-func NewEngine(compiler *ast.Compiler, store storage.Store, trace bool) *Engine {
-	return &Engine{
-		Compiler: compiler,
-		Store:    store,
-		Trace:    trace,
+// Namespaces returns all of the namespaces in the Engine.
+func (e *Engine) Namespaces() []string {
+	var namespaces []string
+	for _, module := range e.Modules() {
+		namespace := strings.Replace(module.Package.Path.String(), "data.", "", 1)
+		if contains(namespaces, namespace) {
+			continue
+		}
+
+		namespaces = append(namespaces, namespace)
+	}
+
+	return namespaces
+}
+
+// Documents returns all of the documents loaded into the engine.
+func (e *Engine) Documents() map[string]string {
+	documents := make(map[string]string)
+	for path, content := range e.result.Documents {
+		documents[path] = fmt.Sprintf("%v", content)
+	}
+
+	return documents
+}
+
+// Policies returns all of the policies loaded into the engine.
+func (e *Engine) Policies() map[string]string {
+	policies := make(map[string]string)
+	for m := range e.result.Modules {
+		policies[e.result.Modules[m].Name] = string(e.result.Modules[m].Raw)
 	}
+
+	return policies
+}
+
+// Compiler returns the compiler from the loaded policies.
+func (e *Engine) Compiler() *ast.Compiler {
+	return e.compiler
+}
+
+// Store returns the store from the loaded documents.
+func (e *Engine) Store() storage.Store {
+	return e.store
+}
+
+// Modules returns the modules from the loaded policies.
+func (e *Engine) Modules() map[string]*ast.Module {
+	return e.result.ParsedModules()
+}
+
+// Runtime returns the runtime of the engine.
+func (e *Engine) Runtime() *ast.Term {
+	env := ast.NewObject()
+	for _, pair := range os.Environ() {
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) == 1 {
+			env.Insert(ast.StringTerm(parts[0]), ast.NullTerm())
+		} else if len(parts) > 1 {
+			env.Insert(ast.StringTerm(parts[0]), ast.StringTerm(parts[1]))
+		}
+	}
+
+	obj := ast.NewObject()
+	obj.Insert(ast.StringTerm("env"), ast.NewTerm(env))
+	obj.Insert(ast.StringTerm("version"), ast.StringTerm(version.Version))
+	obj.Insert(ast.StringTerm("commit"), ast.StringTerm(version.Vcs))
+
+	return ast.NewTerm(obj)
 }
 
 // Query the policy engine with the given query and given input.
 func (e *Engine) Query(ctx context.Context, query string, input interface{}) ([]output.Result, []output.Result, error) {
-	rego, stdout := e.buildRego(e.Trace, query, input)
-	resultSet, err := rego.Eval(ctx)
+	var regoObj *rego.Rego
+	var regoFunc []func(r *rego.Rego)
+	stdout := topdown.NewBufferTracer()
```

Comment:
Nitpick: would it not make sense to move this inside the if statement?

Reply:
yeah, I agree. Hopefully I can get this to just go away!